packages/ubuntu_wsl_setup/lib/services/tcp_helpers.dart

```diff
@@ -0,0 +1,60 @@
+import 'dart:async';
+import 'dart:io';
+import 'dart:math';
+
+// This is inherently a race condition on its own, because in between finding an
+// available port and using it, the system may have binded other socket.
+// Thus, the solution is to hold the socket binded.
+FutureOr<ServerSocket?> _tryBindUnsafe({
+  required int lower,
+  required int higher,
+}) async {
+  var rand = Random.secure();
+  var port = lower + rand.nextInt(higher - lower);
+  try {
+    final socket = await ServerSocket.bind(InternetAddress.loopbackIPv4, port);
+    return socket;
+  } on SocketException {
+    return null;
+  }
+}
+
+class TcpHelperService {
+  /// Ephemeral port ranges (* chart is not on scale)
+  ///
+  /// ```
+  /// <Linux 2.4+>           <Windows 8+>           </Linux 2.4+> </Windows 8+>
+  ///       |----------------------|--------------------|-------------|
+  ///     32768                  49152                60999         65535
+  ///                        [kMinPortNo]           [kMaxPortNo]
+  /// ```
+  static const kMinPortNo = 49152;
+  static const kMaxPortNo = 60999;
+
+  /// Returns a ServerSocket already binded to a randomly-selected port in the
+  /// open interval [lower]-[higher] of the loopback interface where the range
+  /// is a subset of the platform ephemeral port range, limited by [kMinPortNo]
+  /// and [kMaxPortNo]. [maxAttempts] limits the amount of asynchronous
+  /// iterations performed. On failure returns null.
+  static Future<ServerSocket?> getRandomPortSocket({
```

Comment:
this needs to be non-static to be mockable. for example, we could eventually have something like this:
```dart
void main() {
  ...
  registerService<TcpSocketService>(TcpSocketService.new);
  ...
}

void somewhereElse() {
  final service = getService<TcpSocketService>();
  final socket = service.getRandomPortSocket();
  ...
}
```

and in tests:
```dart
final socket = MockServerSocket();
when(socket.port).thenReturn(54321);

final service = MockTcpSocketService();
when(service.getRandomPortSocket()).thenAnswer((_) async => socket);

registerMockService(service);
...
```
this way the tests would be 100% deterministic because they won't try to bind real sockets that could fail for unrelated reasons.